### ISSUE TITLE
3. Complex-Valued Associative Memory with Phase-Coded Storage

### ISSUE THREAD

**Issue by ricardoamartinez** [ricardoamartinez/koopman-physics-neural-cellular-automata#4]
# Complex-Valued Associative Memory Module

## Overview
Implement a **differentiable associative memory** using complex-valued representations for rich pattern storage capacity. This module stores and recalls latent patterns, utilizing complex phase relationships for enhanced interference patterns and memory capacity beyond traditional real-valued approaches.

### Core Memory Architecture
Create a PyTorch module that maintains complex-valued memory patterns with content-addressable retrieval:

```python
class ComplexMemory(nn.Module):
    def __init__(self, memory_size=10, vector_dim=128):
        super(ComplexMemory, self).__init__()
        # Initialize memory matrix with complex values
        # Use real and imag parts from normal distribution
        real = torch.randn(memory_size, vector_dim)
        imag = torch.randn(memory_size, vector_dim)
        self.memory = nn.Parameter(torch.view_as_complex(torch.stack((real, imag), dim=-1)))
        # The memory is a trainable parameter matrix of shape [memory_size, vector_dim] in complex form
    
    def forward(self, query):
        # query: complex tensor of shape [vector_dim] or [batch, vector_dim]
        # Ensure query is complex
        if not torch.is_complex(query):
            # If query given as real tensor, convert to complex by treating it as real part
            query = query.to(self.memory.dtype) + 0j
        
        # Compute similarity between query and each memory item (complex inner product)
        # For batch queries, do matrix multiplication
        # Conjugate the memory for proper complex inner product
        similarity = torch.matmul(query, self.memory.conj().T)  # shape: [batch, memory_size] complex
        
        # Use magnitude of similarity to decide weights (soft retrieval)
        weights = F.softmax(similarity.abs(), dim=-1)           # shape: [batch, memory_size] real weights
        
        # Reconstruct by weighted sum of memory patterns
        recalled = torch.matmul(weights, self.memory)           # shape: [batch, vector_dim] complex
        return recalled
```

### Integration with Agent Loop
Show how memory integrates with NCA latent updates:

```python
# Inside agent's step update (not necessarily within NCADynamics, could be in Agent class)
latent = nca(latent)  # update latent via NCA+Koopman+wavelet

# Derive a query from latent for memory:
query_vec = latent.view(latent.shape[0], -1)      # flatten entire latent grid per batch
# (Optionally reduce dimension via a linear layer to match memory vector_dim)
query_vec = query_vec @ W_q # if needed, where W_q is a learned projection matrix of size [C*H*W, D]
mem_out = memory_module(query_vec)                # complex retrieval, shape [B, D] complex

# Integrate memory output back into latent:
# One approach: add the real part back to latent (assuming latent is real-valued)
mem_out_real = torch.view_as_real(mem_out)        # get tensor of shape [B, D, 2] representing real+imag
mem_out_real = mem_out_real[...,0]                # take real part of recalled vector
# Expand or reshape this to the latent shape for integration
mem_out_expanded = mem_out_real.view(B, C, H, W)
latent = latent + mem_out_expanded               # add retrieved pattern to latent state (residual addition)
```

### 1. Complex Tensor Management
- Initialize memory using `torch.view_as_complex()` with real and imaginary components
- Ensure `nn.Parameter` registration for gradient updates during training
- Use `torch.cfloat` dtype for complex operations
- Handle device placement: `memory_module = ComplexMemory().to(device)`

### 2. Content-Addressable Retrieval
- Implement complex inner product: `torch.matmul(query, memory.conj().T)`
- **Key insight**: Conjugate transpose enables proper complex similarity computation
- Use magnitude (`similarity.abs()`) for attention weights
- Apply softmax for normalized retrieval weights

### 3. Phase-Coded Pattern Storage
- **Real component**: Stores primary pattern information
- **Imaginary component**: Stores phase relationships and secondary features  
- **Phase alignment**: Patterns with aligned phases have higher similarity
- **Interference patterns**: Complex superposition enables richer pattern interactions

#### Strategy A: Residual Addition
```python
# Add real part of recalled memory to latent
mem_real = torch.view_as_real(recalled)[..., 0]  # extract real component
latent = latent + mem_real.view_as(latent)       # residual connection
```

#### Strategy B: Channel Concatenation  
```python
# Concatenate memory as additional channels
mem_channels = recalled.view(B, memory_channels, H, W)
latent = torch.cat([latent, mem_channels], dim=1)  # expand channel dimension
```

#### Strategy C: Gated Integration
```python
# Use learned gating to control memory influence
gate = torch.sigmoid(gate_network(latent))
latent = gate * latent + (1 - gate) * recalled.view_as(latent)
```

### 5. Memory Capacity and Dynamics
- **Hopfield-style storage**: Each pattern stored as outer product component
- **Capacity scaling**: Theoretical capacity ~15% of `vector_dim` for reliable recall
- **Learning dynamics**: Memory patterns updated via gradient descent during training
- **Pattern completion**: Noisy queries retrieve closest stored patterns

### Orthonormal Memory Projection
Ensure stored patterns remain orthogonal for maximum capacity:

```python
def orthonormalize_memory(self):
    """Apply Gram-Schmidt orthonormalization to memory patterns"""
    with torch.no_grad():
        Q, R = torch.linalg.qr(self.memory.T)  # QR decomposition
        self.memory.data = Q.T  # Orthonormal rows
```

### Temporal Memory Updates
Implement Hebbian-style learning for pattern storage:

```python
def update_memory(self, new_pattern, learning_rate=0.01):
    """Update memory with new pattern using Hebbian rule"""
    pattern_normalized = F.normalize(new_pattern.flatten(), dim=0)
    if not torch.is_complex(pattern_normalized):
        pattern_normalized = pattern_normalized + 0j
    
    # Find least-used memory slot or oldest pattern
    similarities = torch.abs(torch.matmul(pattern_normalized.conj(), self.memory.T))
    min_sim_idx = torch.argmin(similarities)
    
    # Update memory slot
    with torch.no_grad():
        self.memory[min_sim_idx] = pattern_normalized
```

## Acceptance Criteria
- [ ] Memory stores complex patterns of shape `[memory_size, vector_dim]`
- [ ] Complex inner product similarity computation works correctly
- [ ] Retrieval produces weighted superposition of stored patterns
- [ ] Integration with NCA preserves gradient flow
- [ ] Memory patterns can be updated during training
- [ ] Phase information properly utilized in similarity computation
- [ ] Performance: retrieval operation <5ms for memory_size=50, vector_dim=128

### Goal State Storage
```python
# Store goal configurations for navigation tasks
goal_latent = encode_goal_state(target_position)
memory_module.update_memory(goal_latent)
```

### Experience Replay
```python
# Store successful action sequences
success_pattern = encode_trajectory(successful_actions)
memory_module.update_memory(success_pattern)
```

### Prototype Learning
```python
# Store category prototypes for classification
for category, examples in training_data:
    prototype = torch.mean(torch.stack(examples), dim=0)
    memory_module.update_memory(prototype)
```

## Integration Notes
The memory module will be instantiated in the main script:
```python
memory_module = ComplexMemory(memory_size=50, vector_dim=latent_dim).to(device)
optimizer = torch.optim.Adam(list(nca.parameters()) + list(memory_module.parameters()), lr=1e-3)
```

Called within the agent loop:
```python
# Query memory after each NCA update
mem_output = memory_module(flatten_latent(latent))
latent = integrate_memory_output(latent, mem_output)
```

This complex-valued associative memory enables the system to **retain information over long timescales** and **recognize familiar latent configurations**, crucial for active-inference agents that must remember goals and past observations. The complex phase dimension potentially doubles effective storage capacity compared to real-valued alternatives.
